perf-event/src/lib.rs

```diff
@@ -815,6 +1076,178 @@ impl IntoRawFd for Counter {
+            unsafe {
+                atomic_store(
+                    raw_field!(page, perf_event_mmap_page, data_tail),
+                    tail + (header.size as u64),
+                    Ordering::Release,
+                )
+            };
+        });
+
+        let config = ParseConfig::from(&self.attrs);
+        Some(Record::parse(&config, &header, &mut reader))
+    }
+
+    /// Read the next record from the ring buffer. This method will block (with
+    /// an optional timeout) until a new record is available.
+    ///
+    /// If this sampler is only enabled for a single process and that process
+    /// exits, this method will return `None` even if no timeout is passed.
+    /// Note that this only works on Linux 3.18 and above.
+    pub fn next_blocking(&mut self, timeout: Option<Duration>) -> Option<samples::Record> {
+        let deadline = timeout.map(|timeout| Instant::now() + timeout);
+
+        loop {
+            if let Some(record) = self.next() {
+                return Some(record);
+            }
+
+            let timeout = match deadline {
+                Some(deadline) => deadline
+                    .checked_duration_since(Instant::now())?
+                    .as_millis()
+                    .min(libc::c_int::MAX as u128) as libc::c_int,
+                None => -1,
+            };
+
+            let mut pollfd = libc::pollfd {
+                fd: self.as_raw_fd(),
+                events: libc::POLLIN,
+                revents: 0,
+            };
+
+            match check_errno_syscall(|| unsafe { libc::poll(&mut pollfd, 1, timeout) }) {
+                // poll timed out.
+                Ok(0) => return None,
+                // The sampler was tracking a single other process and that
+                // process has exited.
+                Ok(_) if pollfd.revents & libc::POLLHUP != 0 => return None,
+                // Must be POLLIN, there should be an event ready.
+                Ok(_) => continue,
+                Err(e) => match e.raw_os_error() {
+                    Some(libc::EINTR) => continue,
+                    // The only other possible kernel errors here are so rare
+                    // that it doesn't make sense to make this API have a
+                    // result because of them. To whit, they are:
+                    // - EINVAL - the process ran out of file descriptors
+                    // - ENOMEM - the kernel couldn't allocate memory for the
+                    //            poll datastructures.
+                    _ => return None,
+                },
+            }
+        }
+    }
+
+    fn page(&self) -> *const sys::bindings::perf_event_mmap_page {
+        self.mmap.as_ptr() as *const _
+    }
+}
+
+impl std::ops::Deref for Sampler {
+    type Target = Counter;
+
+    fn deref(&self) -> &Self::Target {
+        &self.counter
+    }
+}
+
+impl std::ops::DerefMut for Sampler {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.counter
+    }
+}
```

Comment:
This way of making `Sampler` support the operations of `Counter` is considered an anti-pattern in Rust:
https://rust-unofficial.github.io/patterns/anti_patterns/deref.html

Possible alternatives:
- Add a `counter` method to `Sampler`, and let people call that to get the underlying `Counter`.
- Create a trait with the methods, and have both `Sampler` and `Counter` implement that.

Reply:
Unfortunately, both of those alternatives aren't great :(

As a third alternative:
- Copy the methods from `Counter` to `Sampler` and have them delegate to the real implementations in `Counter`

I think that either keeping the `Deref` impls (despite it being an antipattern) or copying the methods is the best choice. Ultimately, though, the choice is yours and I'll implement whichever one you prefer.

Reply:
Just thinking this over: Some of the methods only make sense for counted events. Not all sampled events are counted events. But there are counters that record their overflows to the mmap area, so certainly `Sampler` needs the counter-related methods. So I guess there's nothing that can be omitted.

Let's just copy the methods for now.

Reply:
I've removed the deref impls and used a macro to add the methods to both `Counter` and `Sampler`.